geopandas/tools/sjoin.py

```diff
@@ -186,6 +180,351 @@ def find_contains(a1, a2):
+          to speed up execution.
+    nearest_distances: bool, default False
+        If True, report the distance for each match in new column
+          named "nearest_distances".
+    """
+    # ------------------------------ CHECK INPUTS ------------------------------
+    allowed_hows = ["left", "right", "inner"]
+    if how not in allowed_hows:
+        raise ValueError(
+            '`how` was "%s" but is expected to be in %s' % (how, allowed_hows)
+        )
+
+    _basic_checks(left_df, right_df)
+
+    # ----------------- RENAME INDEXES FOR RTREE COMPATIBILITY -----------------
+    (
+        left_df,
+        right_df,
+        left_index_name,
+        right_index_name,
+        index_left,
+        index_right,
+    ) = _rename_indexes(left_df, right_df, lsuffix, rsuffix)
+
+    # ------------------------ COMPUTE SPATIAL JOIN ----------------------------
+
+    # get spatial index
+    tree_idx = right_df.sindex
+
+    # validate max_search_neighbors and search_radius params
+    if max_search_neighbors is not None:
+        assert max_search_neighbors >= 1 and isinstance(
+            max_search_neighbors, int
+        ), "max_search_neighbors must be an integer and >= 1"
+        # warn about using rtree < 0.9.4 and max_search_neighbors option
+        # see https://github.com/Toblerity/rtree/pull/141
+
+        if [int(i) for i in RTREE_VERSION.split(".")] < [0, 9, 4]:
+            warn(
+                "Using an rtree version < 0.9.4 may cause inconsistent "
+                "results when using max_search_neighbors. Consider using a "
+                "large number or max_search_neighbors=None."
+            )
+    if search_radius is not None:
+        assert search_radius >= 0, "search_radius must be >=0"
+
+    def query_index(geo_in_l):
+        """
+        Queries the spatial index to filter results.
+        """
+        # restric by radius
+        if search_radius is not None:
+            if geo_in_l.geom_type == "Point" and search_radius == 0:
+                # applying .buffer(0) to a point returns an empty geometry
+                geo_search = geo_in_l.coords[0]
+            else:  # polygon or buffer > 0
+                geo_search = geo_in_l.buffer(search_radius).bounds
+            in_radius = list(tree_idx.intersection(geo_search))
+        else:
+            in_radius = right_df.index
+        # find neighbors
+        if max_search_neighbors is not None:
+            neighbors = list(
+                tree_idx.nearest(geo_in_l.bounds, num_results=max_search_neighbors)
+            )
+        else:
+            neighbors = right_df.index
+        # ensure indexes are int
+        return (int(idx) for idx in neighbors if idx in in_radius)
+
+    # the accuracy of the spatial index is limited, so we need to manually
+    # check each set of matches for actual distance
+    l_idx = []
+    r_idx = []
+    distances = []
+    for ind_in_left, geo_in_l in enumerate(left_df.geometry):
```

Comment:
Better would be to use .iteritems() within enumerate. It should be faster loop if I am not mistaken.

Reply:
So:
```python3
for ind_in_left, geo_in_l in left_df.geometry.iteritems():
```
Is that correct?

Reply:
No, it should be 

```py
for ind_in_left, (_, geo_in_l) in enumerate(left_df.geometry.iteritems()):
```

But I am not sure if it is actually faster in this case. You need enumerate as we use row number in the spatial index, not the index value.

Reply:
It seems that the current way is actually faster:
```python3
import numpy as np
import pandas as pd
import time

y = np.random.rand(10000, 2)
df = pd.DataFrame({'y1': y[:, 0], 'y2': y[:, 1]})


t = time.time()
for ind_in_left, (_, geo_in_l) in enumerate(df.y1.iteritems()):
  pass
t1 = time.time() - t

t = time.time()
for ind_in_left, geo_in_l in enumerate(df.y1):
  pass
t2 = time.time() - t

print(t1, t2)
```

```
0.00573420524597168 0.0026712417602539062
```

Reply:
Yes, agree. Keep it. 
<img width="657" alt="Screenshot 2020-03-03 23 08 27" src="https://user-images.githubusercontent.com/36797143/75828542-0221a300-5da4-11ea-9191-b5fa18c480a5.png">